embed/serve.go

```diff
@@ -102,22 +102,38 @@ func (sctx *serveCtx) serve(
 	servElection := v3election.NewElectionServer(v3c)
 	servLock := v3lock.NewLockServer(v3c)
 
-	var gs *grpc.Server
-	defer func() {
-		if err != nil && gs != nil {
-			gs.Stop()
-		}
-	}()
-
 	if sctx.insecure {
-		gs = v3rpc.Server(s, nil, gopts...)
+		gs := v3rpc.Server(s, nil, gopts...)
 		v3electionpb.RegisterElectionServer(gs, servElection)
 		v3lockpb.RegisterLockServer(gs, servLock)
 		if sctx.serviceRegister != nil {
 			sctx.serviceRegister(gs)
 		}
+
+		defer func(gs *grpc.Server) {
+			if err == nil {
+				return
+			}
```

Comment:
The gs will not be stopped in such case?

Reply:
`gs` is serving in the background. So I guess on etcd stop, gs will be gracefully stopped.

Reply:
> The gs will not be stopped in such case?

Currently, it won't happen. The `cMux.Serve` always return error. I was thinking that we can remove that error check and just stop it. But this pr is used to fix data-race issue. It can be updated in the follow-up.

Reply:
OK, thx for the clarification.